### PR TITLE
feat(form-workspaces-be/2): add get workspaces functionality

### PR DESCRIPTION
**Pull request by hanstirtaputra**
Part of  #4039 

## Solution
Add GET `api/v3/admin/workspaces` endpoint business logic

**Breaking Changes** 
- [x] No - this PR is backwards compatible  

## Tests
Unit tests are included for this endpoint

### CODE REVIEW

File: src/app/modules/workspace/__tests__/workspace.routes.spec.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    await dbHandler.clearDatabase()
+    jest.restoreAllMocks()
+  })
+  afterAll(async () => await dbHandler.closeDatabase())
+
+  describe('GET /workspaces', () => {
+    const GET_WORKSPACES_ENDPOINT = '/workspaces'
+
+    it('should return 200 with an empty array when a user has no workspaces', async () => {
+      const response = await request.get('/workspaces')
```
Comment (karrui) [inline src/app/modules/workspace/__tests__/workspace.routes.spec.ts:54]: nit: should this use the `GET_WORKSPACES_ENDPOINT` constant too?

File: src/app/modules/workspace/__tests__/workspace.routes.spec.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+const WorkspaceModel = getWorkspaceModel(mongoose)
+
+const app = setupApp('/workspaces', WorkspacesRouter, {
+  setupWithAuth: true,
+})
+
+const MOCK_USER_ID = new ObjectId()
+const MOCK_FORM_ID = new ObjectId()
+const MOCK_WORKSPACE_ID = new ObjectId()
+const MOCK_WORKSPACE_FIELDS = {
```
Comment (karrui) [inline src/app/modules/workspace/__tests__/workspace.routes.spec.ts]: nit: should this be called `MOCK_WORKSPACE_DOC` instead?

File: src/app/modules/workspace/__tests__/workspace.routes.spec.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      const response = await request.get('/workspaces')
+
+      expect(response.status).toEqual(200)
+      expect(response.body).toEqual([])
+    })
+
+    it("should return 200 with an array of the user's workspaces", async () => {
+      await WorkspaceModel.create(MOCK_WORKSPACE_FIELDS)
+      const response = await request.get(GET_WORKSPACES_ENDPOINT)
+      const expected = await WorkspaceModel.find({ _id: MOCK_WORKSPACE_ID })
```
Comment (karrui) [inline src/app/modules/workspace/__tests__/workspace.routes.spec.ts]: nit: would it be clearer for the reader if we used `MOCK_WORKSPACE_DOC._id` (even though they are the same thing, haha)

**Review by karrui — APPROVED** [review]
lgtm

File: shared/types/workspace.ts [hunk 4]
```diff
@@ -5,7 +5,6 @@ import { UserId } from './user'
 export type WorkspaceId = Opaque<string, 'WorkspaceId'>
 
 export type Workspace = {
-  _id: WorkspaceId
```
Comment (mantariksh) [inline shared/types/workspace.ts:8]: why remove this?

File: src/app/modules/workspace/__tests__/workspace.routes.spec.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    })
+    request = await createAuthedSession(user.email, request)
+  })
+  afterEach(async () => {
+    await dbHandler.clearDatabase()
+    jest.restoreAllMocks()
+  })
+  afterAll(async () => await dbHandler.closeDatabase())
+
+  describe('GET /workspaces', () => {
```
Comment (mantariksh) [inline src/app/modules/workspace/__tests__/workspace.routes.spec.ts:50]: should we test the ordering of workspaces as well?

**Review by mantariksh — COMMENTED** [review]
do we need model-level tests for `WorkspaceSchema.statics.getWorkspaces`?

File: src/app/models/workspace.server.model.ts [hunk 6]
```diff
@@ -40,6 +41,16 @@ const compileWorkspaceModel = (db: Mongoose): IWorkspaceModel => {
     admin: 1,
   })
 
+  WorkspaceSchema.virtual('count').get(function (this: IWorkspaceSchema) {
```
Comment (mantariksh) [inline src/app/models/workspace.server.model.ts]: btw, cool use of virtuals! this is a good use case for it (although I'm not sure what it's used for? but I'll probably find out in later PRs)

File: src/app/modules/workspace/workspace.utils.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import { ErrorResponseData } from '../core/core.types'
+
+const logger = createLoggerWithLabel(module)
+
+export const mapRouteError = (
+  error: ApplicationError,
+  coreErrorMessage?: string,
+): ErrorResponseData => {
+  switch (error.constructor) {
+    case DatabaseError:
```
Comment (mantariksh) [inline src/app/modules/workspace/workspace.utils.ts:22]: what about the other specific error classes returned by `transformMongoError`, like `DatabaseValidationError` or `DatabaseConflictError`? should those return different status codes?

Comment (hanstirtaputra) [reply src/app/modules/workspace/workspace.utils.ts:22]: Will update to include the other MongoErrors. Yes I think those should return different status codes

File: src/app/modules/workspace/__tests__/workspace.routes.spec.ts (hunk 2 above)
Comment (hanstirtaputra) [reply src/app/modules/workspace/__tests__/workspace.routes.spec.ts]: Yes it should, thanks

File: src/app/modules/workspace/__tests__/workspace.routes.spec.ts (hunk 3 above)
Comment (hanstirtaputra) [reply src/app/modules/workspace/__tests__/workspace.routes.spec.ts]: Yep, I think so, will update to use `MOCK_WORKSPACE_DOC._id` instead

File: shared/types/workspace.ts (hunk 4 above)
Comment (hanstirtaputra) [reply shared/types/workspace.ts:8]: Thought this was the wrong type to add the `_id` in. I've reverted this change

File: src/app/modules/workspace/__tests__/workspace.routes.spec.ts (hunk 5 above)
Comment (hanstirtaputra) [reply src/app/modules/workspace/__tests__/workspace.routes.spec.ts:50]: Yep, I've added test for workspace ordering as well

File: src/app/models/workspace.server.model.ts (hunk 6 above)
Comment (hanstirtaputra) [reply src/app/models/workspace.server.model.ts]: It's used in the FE, it's not necessary as clients can just use the `formIds.length`, but I felt this was just a nice-to-have

Comment (mantariksh) [reply src/app/models/workspace.server.model.ts]: I would say if it's not necessary then let's not add it in. creating derived properties makes it harder to keep everything in sync, and this doesn't give us any performance boost

Comment (hanstirtaputra) [reply src/app/models/workspace.server.model.ts]: Okay then, I've updated the PR to remove the `count` virtual property!

File: src/app/modules/workspace/__tests__/workspace.routes.spec.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+    it('should return 401 when user is not logged in', async () => {
+      await logoutSession(request)
+      const response = await request.get(GET_WORKSPACES_ENDPOINT)
+
+      expect(response.status).toEqual(401)
+      expect(response.body).toEqual({ message: 'User is unauthorized.' })
+    })
+
+    it('should return 500 when database errors occur', async () => {
```
Comment (mantariksh) [inline src/app/modules/workspace/__tests__/workspace.routes.spec.ts:100]: we can test for the other types of database errors as well. probably sufficient to do it just in integration tests, no need to add them in every other layer

File: src/app/routes/api/v3/admin/workspaces/workspaces.routes.ts [hunk 9]
```diff
@@ -24,7 +24,6 @@ WorkspacesRouter.route('/')
    *
    * @returns 200 with a list of workspaces owned by the user
    * @returns 401 when user is not logged in
```
Comment (mantariksh) [inline src/app/routes/api/v3/admin/workspaces/workspaces.routes.ts:26]: can update this with the new status codes

**Review by mantariksh — APPROVED** [review]
couple of more nits, otherwise lgtm!

File: src/app/modules/workspace/__tests__/workspace.routes.spec.ts (hunk 8 above)
Comment (hanstirtaputra) [reply src/app/modules/workspace/__tests__/workspace.routes.spec.ts:100]: Similar to the other testcases in the codebase, I added the test to the controller layer as we can mock the `DatabaseConflictError` there, whereas in the integration test it's harder to mock the difference in mongoose versions

**Pull request merged**